### DISCUSSION TITLE
How to correctly buffer convex polygon?

### DISCUSSION THREAD

**Discussion by xakod** [boostorg/geometry#1019] [Q&A] [answered]
Hi, I wanna to buffer a polygon.
I have a polygon containing 12 points so I want that buffered polygon also contains 12 points. But my buffered polygon contains 33 points. Why?
Gist https://gist.github.com/Jihadist/864acc5efefd26c79fbb262ace5c5604
![buffer_polygon](https://user-images.githubusercontent.com/26579016/174589246-fb5da531-56b6-4566-9bb4-73c0f09f482e.png)
Image looks like ok

I tried boost::geometry::strategy::buffer::point_square  and point_circle but there is no difference. What I do wrong?
Thanx in advance!

**Comment by awulkiew** [comment, marked as answer]
AFAIU the number of points of a buffer is not guaranteed. `point_xxx` strategies are used for buffer of a single point and multi-points so they won't help here. Join strategies are used at vertices so I think you need to use `boost::geometry::strategy::buffer::join_miter` instead of `join_round`. If there are still too many points in the result you could e.g. use `simplify()` to get rid of the points lying on straight parts.

**Reply by awulkiew** [reply]
I just saw your code and you're already using mitter strategy. I have no other ideas than the use of `simplify()`. Could you write WKTs of input and output here for reference? e.g. with `std::cout << bg::wkt(mp);`

@barendgehrels do you have an idea?

**Reply by xakod** [reply]
I thought about simplify but I don't know max_distance  and have no ideas how to determine it for entire polygon.

Input corrected
```POLYGON((577255 928582,577255 928582,577228 928786,577245 932654,577619 933122,580589 933287,580929 933297,583237 932957,583504 932546,583652 929953,582964 928631,577255 928582))``` 
Output mp
```MULTIPOLYGON(((576958 928543,576931 928747,576928 928767,576928 928787,576945 932655,576945 932760,577011 932841,577385 933309,577468 933414,577602 933422,580572 933587,580576 933587,580580 933587,580920 933597,580947 933598,580973 933594,583281 933254,583415 933234,583489 933120,583756 932709,583799 932643,583804 932563,583952 929970,583956 929888,583918 929815,583230 928493,583147 928333,582967 928331,577258 928282,576992 928280,576958 928543)))```

**Reply by vissarion** [reply]
`max_distance` is the maximum [Hausdorff distance](https://en.wikipedia.org/wiki/Hausdorff_distance) of the two polygons (input and simplified). Unfortunately, this is a input parameter for the algorithm implemented in Boost Geometry. So you either go by trial and error (a small fraction of the buffer distance could work) or you compute in brute-force all the distances of the output points to the polygon (by excluding from the polygon the point that you compute the distance against) and then find the right distance that keeps the desired number of points after simplification. It is possible to find something more clever that I cannot think of right now.

**Reply by xakod** [reply]
I found another approach but I'm not sure that it's correct
```c++
   // mp is from gist - buffered multi_polygon
    int prev_degree = std::numeric_limits<int>::max();
    {
        const auto& poly = mp.front().outer();
        decltype(mp)::value_type filtered_poly;
        for (auto it = poly.begin(); it != poly.end(); it++) {
            decltype(mp)::value_type::point_type start, end;
            if (it != poly.end() - 1) {
                start = *it;
                end = *(it + 1);
            } else {
                start = *it;
                end = *poly.begin();
            }
            auto azimuth = bg::azimuth(start, end);

            boost::units::quantity<
                boost::units::unit<
                    boost::units::plane_angle_dimension,
                    boost::units::si::system>>
                angle_radians(azimuth * boost::units::si::radian);

            auto bdegree = static_cast<boost::units::quantity<
                boost::units::unit<
                    boost::units::plane_angle_dimension,
                    boost::units::degree::system>>>(angle_radians);
            int current_degree = static_cast<int>(bdegree.value());

            if (prev_degree != current_degree) {
                filtered_poly.outer().push_back(start);
            }

            prev_degree = current_degree;
        }

        std::cout << "Filtered size: " << filtered_poly.outer().size() << std::endl;

        std::cout << bg::wkt(filtered_poly) << std::endl;
    }
```

mp internal size: 31
Filtered size: 12
```
POLYGON((576958 928543,576928 928767,576945 932760,577468 933414,580576 933587,580947 933598,583415 933234,583799 932643,583956 929888,583147 928333,576992 928280,576958 928543))
```

It creates from buffered 31 points polygon 12 points polygon (it's the same size as input polygon size). 
Image looks like ok
![buffer_polygon_2](https://user-images.githubusercontent.com/26579016/174766418-b79664e6-be74-40ed-a9ae-2c4d7998f4bb.png)

**Reply by barendgehrels** [reply]
This manual filtering looks correct and is about the same as what simplify would do.

If you want to simplify redundant points, take a very small distance (for example 0.001 for your case). The simplify distance not (not at all) related to the distance of the buffer. For simplifying only redundant points, always use a very small distance.

If I use that, I get also 11 points, so even less than you started with. Because there is probably one redundant point in your input.
`issue_1019 input 12 -> 31 simplified 11`

Internally it creates this:
![image](https://user-images.githubusercontent.com/334849/174988471-28c72d01-4b01-471b-a876-d441315ce9a6.png)

So that is also where the extra points come from. It should be possible to omit them - but this was never asked before and kind of forgotten. For rounded angles it will not make much difference.

**Reply by barendgehrels** [reply]
BTW: having the same number of input/output points was never a requirement for us, because it only makes sense for convex polygons, with miters. As soon as there are concavities the shape might become different in many ways, for example:

![image](https://user-images.githubusercontent.com/334849/174990946-869fd71c-e9a5-4368-9ee9-ea03fe79262b.png)

So it makes no sense to check that internally.
But the redundant points could indeed be filtered out.

**Reply by xakod** [reply]
First of all, thank you for this lection! I use only convex polygons so it reasonable for my but now i understand that we can another polygons. 
Eventually, can I use this simplifing coefficient as a general coeff appropriate for smaller or bigger polygons than my?

**Reply by awulkiew** [reply]
If you're only dealing with convex polygons, need only miter joins and want to preserve the number of points then maybe you could write your own algorithm?

You'd have to iterate over the segments of a polygon treating them as 2D vectors and calculate the third vector between them. This would be a vector between bisection point which is half of the sum of the vectors `(s1+ s2) * 0.5` and the point between the segments. Segments having the same direction need special care to avoid generating a vector having `length = 0`. Then you only need to normalize and calculate the length of this vector based on the angle between the segments. This algorithm would have complexity O(N) so lesser than `buffer()` and should be a lot faster.

AFAIU this is what `join_mitter` [1] does so maybe you could simply use this strategy directly and since it generates three points of a join you could only use the middle one. Though it takes parameters that are not intuitive for me so you have to ask @barendgehrels how to use it.

[1] https://github.com/boostorg/geometry/blob/develop/include/boost/geometry/strategies/cartesian/buffer_join_miter.hpp#L52

**Reply by barendgehrels** [reply]
> First of all, thank you for this lection! I use only convex polygons so it reasonable for my but now i understand that we can another polygons. Eventually, can I use this simplifing coefficient as a general coeff appropriate for smaller or bigger polygons than my?

Yes, this simplifying distance you can use in general. If it's near zero, it just removes redundant points. You can even try 0.0 too. This page describes the simplify algorithm in general https://en.wikipedia.org/wiki/Ramer%E2%80%93Douglas%E2%80%93Peucker_algorithm (here called epsilon

**Reply by barendgehrels** [reply]
> If you're only dealing with convex polygons, need only miter joins and want to preserve the number of points then maybe you could write your own algorithm?
> 

It's possible of course - but using buffer itself should just work as desired of course. Maybe we should move this to an `issue` and remove the redundant points. It will not much extra code.

**Reply by vissarion** [reply]
A side note (for clarity) on the connection between simplify distance and buffer distance. To my understanding they are connected as follows: the distance of a point created by the buffer (the points on the circular arc in OGC standard buffer, miter points should be similar) to the closest segment created if we remove that point is a function of the buffer distance r and the angle $\theta$ (of the circular arc, equivalently the angle of the orthogonal normals of the polygon segments), i.e. $$\sqrt{4r^2\sin^2\frac{\theta}{4}-r^2\sin^2\frac{\theta}{2}}.$$ This is exactly the distance used by simplify algorithm to filter out points from the input. If a point is close to be "redundant" then that distance is very small (since the angle is very small) so the small numbers that @barendgehrels suggests are working. I do not know if this is needed in practice but eventually one can actually precompute those numbers before running simplify.

**Reply by barendgehrels** [reply]
Thanks @vissarion . Now I understand your reasoning a bit better.
But I don't know if I agree. It's not that we want to remove some intermediate points from the generated circle. You can specify the number of points there, so usually there is no need to remove them afterwards. We want to remove the middle point from three collinear points. That point is redundant. For that, you can always use the same simplify distance of nearly zero.

**Reply by vissarion** [reply]
Sure, I agree with you, I just wanted to argue that the is a connection
between buffer distance and simplify distance. Sorry if I made the
discussion more complicated than needed.
